support-workers/src/main/scala/com/gu/support/workers/lambdas/SendThankYouEmail.scala

```diff
@@ -36,12 +37,21 @@ class SendThankYouEmail(
   ): FutureHandlerResult = {
     val emailBuilder = new EmailBuilder(
       services.zuoraService.getMandateIdFromAccountNumber,
+      _ match {
+        case None => Future.successful(None)
+        case Some(n) =>
+          for {
+            response <- services.paperRoundService.agents()
+          } yield response.data.agents.find(_.refId == n)
```

Comment:
what is "n"? the reference about the agent? I'm still not keen on modelling an opaque id as an integer data type as there are too many caveats about overflow/negatives etc.

It might be worth having a trait rather than a function type here, just so it's clearer what's going on (It may have been me that did the original function type in the EmailBuilder, and I now know it's not very readable compared with a trait!)

Reply:
> Thanks for your detailed reply and all good points.
> 
> I think my concerns are also about clear naming and also type checking, if you have a wrapper type it means you can't mix it up with other unrelated values, also it provides documentation/naming for the code. It's a good point that there's a cost too that has to be worth it.
> 
> For the agent ID it makes more sense to have a proper type as it goes throughout the codebase a lot more.

I’m sorry, I’ve been too stubborn about this: I’ll change it to use a wrapper type throughout. Have you a preference for the underlying type? I asked Chris from PaperRound about the range for IDs and he said a 32-bit integer should be fine, but I’m uncertain whether I’m happy to rely on that. I’m leaning towards scala’s`BigInt` at the moment, and wondering how awkward it would be to use a string instead.

> In the case of the function, because it's passed in one place and used in the target function it's not a big issue. I just took a while to go back and forth to work out what it was, I thought it was related to the DD mandate stuff on the line above at first.

Now that it’s named following Rupert’s suggestion I think it’s clearer, so I’d rather stick with it than change it to a trait, if that’s ok?
 
> I think the cases that really confuse me (in other codebases), functions are passed in several levels and by that time it's very hard to navigate to the possible implementations.

This is an interesting point: when it’s a trait you can search for all classes that implement the trait to find all the implementations in a codebase, whereas when it’s functions it’s hard to search for all implementations other than climbing up through function calls?